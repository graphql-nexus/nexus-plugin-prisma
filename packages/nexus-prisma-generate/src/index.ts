#!/usr/bin/env node

import { existsSync, mkdirSync, writeFileSync } from 'fs'
import {
  GraphQLEnumType,
  GraphQLField,
  GraphQLInputField,
  GraphQLInputObjectType,
  GraphQLObjectType,
  GraphQLSchema,
  isEnumType,
  isInputObjectType,
  isObjectType,
  isScalarType,
  introspectionFromSchema,
} from 'graphql'
import * as meow from 'meow'
import { EOL } from 'os'
import { join } from 'path'
import { ISDL } from 'prisma-datamodel'
import { generateCRUDSchemaFromInternalISDL } from 'prisma-generate-schema'
import {
  findDatamodelAndComputeSchema,
  findRootDirectory,
  getImportPathRelativeToOutput,
} from './config'
import { getFinalType, getTypeName, isList, isRequired } from './graphql'

const cli = meow(
  `
    nexus-prisma-generate prisma-client-dir output

    > Generate the TypeScript types for nexus-prisma

    -----
    
    Inputs should be relative to the root of your project

    --client (required): Path to your prisma-client directory (eg: ./generated/prisma-client/)
    --output (required): Path to directory where you want to output the typings (eg: ./generated/nexus-prisma)
    --js     (optional): Whether to generate the types for Javascript
`,
  {
    flags: {
      client: {
        type: 'string',
      },
      output: {
        type: 'string',
      },
      js: {
        type: 'boolean',
        default: false,
      },
    },
  },
)

main(cli)

function main(cli: meow.Result) {
  const { client: prismaClientDir, output, js: jsMode } = cli.flags

  if (!prismaClientDir) {
    console.log('ERROR: Missing argument --client')
    process.exit(1)
  }

  if (!output) {
    console.log('ERROR: Missing argument --output')
    process.exit(1)
  }

  const rootPath = findRootDirectory()
  const resolvedOutput = output.startsWith('/')
    ? output
    : join(rootPath, output)
  const resolvedPrismaClientDir = prismaClientDir.startsWith('/')
    ? prismaClientDir
    : join(rootPath, prismaClientDir)

  if (!existsSync(resolvedPrismaClientDir)) {
    console.log(
      `ERROR: Cannot find --client path at ${resolvedPrismaClientDir}`,
    )
    process.exit(1)
  }

  try {
    // Create the output directories if needed (mkdir -p)
    mkdirSync(resolvedOutput, { recursive: true })
  } catch (e) {
    if (e.code !== 'EEXIST') throw e
  }

  const { datamodel, databaseType } = findDatamodelAndComputeSchema()

  try {
    const schema = generateCRUDSchemaFromInternalISDL(datamodel, databaseType)
    const renderedDatamodel = renderDatamodel(datamodel, schema)
    const nexusPrismaTypesPath = join(rootPath, output, 'nexus-prisma.ts')
    const nexusPrismaTypes = renderNexusPrismaTypes(
      schema,
      getImportPathRelativeToOutput(
        resolvedPrismaClientDir,
        nexusPrismaTypesPath,
      ),
    )

    writeFileSync(nexusPrismaTypesPath, nexusPrismaTypes)

    if (jsMode) {
      const datamodelPath = join(rootPath, output, 'datamodel-info.js')
      const indexPath = join(rootPath, output, 'index.js')

      writeFileSync(datamodelPath, `module.exports = ${renderedDatamodel}`)
      writeFileSync(indexPath, renderIndexJs())
    } else {
      const datamodelPath = join(rootPath, output, 'datamodel-info.ts')
      const indexPath = join(rootPath, output, 'index.ts')

      writeFileSync(datamodelPath, `export default ${renderedDatamodel}`)
      writeFileSync(indexPath, `export { default } from './datamodel-info'`)
    }

    console.log(`Types generated at ${output}`)
  } catch (e) {
    console.error(e)
  }
}

function renderIndexJs() {
  return `\
const datamodelInfo = require('./datamodel-info')
  
module.exports = datamodelInfo
  `
}

function renderDatamodel(datamodel: ISDL, schema: GraphQLSchema) {
  return `\
{
  uniqueFieldsByModel: {
${datamodel.types
  .map(
    type =>
      `    ${type.name}: [${type.fields
        .filter(field => field.isUnique)
        .map(field => `'${field.name}'`)
        .join(', ')}]`,
  )
  .join(',' + EOL)}
  },
  schema: ${JSON.stringify(introspectionFromSchema(schema), null, 2)}
}
  `
}

function renderNexusPrismaTypes(
  schema: GraphQLSchema,
  prismaClientPath: string,
) {
  const types = Object.values(schema.getTypeMap())
  const objectTypes = types.filter(
    t => isObjectType(t) && !t.name.startsWith('__'),
  ) as GraphQLObjectType[]
  const inputTypes = types.filter(isInputObjectType)
  const enumTypes = types.filter(
    t => isEnumType(t) && !t.name.startsWith('__'),
  ) as GraphQLEnumType[]

  return `\
// GENERATED TYPES FOR NEXUS-PRISMA. /!\\ DO NOT EDIT MANUALLY

import { core } from 'nexus'
import { GraphQLResolveInfo } from 'graphql'
import * as prisma from '${prismaClientPath}'

declare global {
  interface NexusPrismaGen extends NexusPrismaTypes {}
}

export interface NexusPrismaTypes {
  objectTypes: {
    fields: {
${objectTypes
  .map(type => `      ${type.name}: ${getObjectTypeFieldsName(type)}`)
  .join(EOL)}
    }
    fieldsDetails: {
${objectTypes
  .map(type => `      ${type.name}: ${getObjectTypeFieldsDetailsName(type)}`)
  .join(EOL)}
    }
  }
  inputTypes: {
    fields: {
${inputTypes
  .map(
    type =>
      `      ${getInputObjectTypeName(type)}: ${getInputObjectTypeFieldsName(
        type,
      )}`,
  )
  .join(EOL)}
    }
  }
  enumTypes: {
${enumTypes
  .map(type => `    ${type.name}: ${getEnumTypeName(type)},`)
  .join(EOL)}
  }
}

${objectTypes.map(renderObjectType).join(EOL)}

${inputTypes.map(renderInputType).join(EOL)}

${enumTypes.map(renderEnumType).join(EOL)}
  `
}

function renderObjectType(type: GraphQLObjectType) {
  const fields = Object.values(type.getFields())

  return `\
// Types for ${type.name}

${renderFields(type, fields)}

${renderFieldsArgs(type, fields)}

${renderTypeFieldDetails(type, fields)}
`
}

function renderTypeFieldDetails(
  type: GraphQLObjectType,
  fields: GraphQLField<any, any>[],
) {
  return `\
export interface ${getObjectTypeFieldsDetailsName(type)} {
${fields
  .map(
    field => `\
  ${field.name}: {
    type: '${getTypeName(field.type)}'
    args: ${
      field.args.length > 0
        ? `Record<${getTypeFieldArgName(
            type,
            field,
          )}, core.NexusArgDef<string>>`
        : '{}'
    }
    description: string
    list: ${isList(field.type) ? true : undefined}
    nullable: ${!isRequired(field.type)}
    resolve: ${
      isScalarType(getFinalType(field.type))
        ? undefined
        : `(
      root: core.RootValue<"${type.name}">,
      args: ${renderResolverArgs(field)},
      context: core.GetGen<"context">,
      info?: GraphQLResolveInfo
    ) => ${renderResolverReturnType(field)}`
    }
  }`,
  )
  .join(EOL)}
}
  `
}

function renderResolverArgs(field: GraphQLField<any, any>) {
  return `{\
 ${field.args
   .map(
     arg => `${arg.name}${isRequired(arg.type) ? '' : '?'}: ${getTSType(arg)}`,
   )
   .join(', ')} }\
  `
}

function renderFields(
  type: GraphQLObjectType,
  fields: GraphQLField<any, any>[],
) {
  return `\
type ${getObjectTypeFieldsName(type)} =
  | ${getExposableFieldsTypeName(type)}
${fields
  .map(
    f =>
      `  | { name: '${f.name}', args?: ${
        f.args.length > 0
          ? `${getTypeFieldArgName(type, f)}[] | false`
          : '[] | false'
      }, alias?: string  } `,
  )
  .join(EOL)}

type ${getExposableFieldsTypeName(type)} =
${fields.map(f => `  | '${f.name}'`).join(EOL)}
`
}

function renderFieldsArgs(
  type: GraphQLObjectType,
  fields: GraphQLField<any, any>[],
) {
  return `\
${fields
  .filter(field => field.args.length > 0)
  .map(field => renderFieldArg(type, field))
  .join(EOL)}
  `
}

function renderFieldArg(
  type: GraphQLObjectType,
  field: GraphQLField<any, any>,
) {
  return `\
type ${getTypeFieldArgName(type, field)} =
${field.args.map(arg => `  | '${arg.name}'`).join(EOL)}`
}

function getTSType(graphqlType: GraphQLField<any, any> | GraphQLInputField) {
  const graphqlToTypescript: Record<string, string> = {
    String: 'string',
    Boolean: 'boolean',
    ID: 'string',
    Int: 'number',
    Float: 'number',
    DateTime: 'string',
  }
  const finalType = getFinalType(graphqlType.type)

  let returnType = ''

  if (isScalarType(finalType)) {
    returnType = graphqlToTypescript[getTypeName(finalType)]
  } else if (isInputObjectType(finalType)) {
    returnType = getInputObjectTypeName(finalType)
  } else {
    returnType = `prisma.${getTypeName(finalType)}`
  }

  if (isList(graphqlType.type)) {
    returnType += '[]'
  }

  if (!isRequired(graphqlType.type)) {
    returnType += ' | null'
  }

  return returnType
}

function renderResolverReturnType(field: GraphQLField<any, any>) {
  const returnType = getTSType(field)

  return `Promise<${returnType}> | ${returnType}`
}

function renderInputType(input: GraphQLInputObjectType) {
  const fields = Object.values(input.getFields())
  return `\
export interface ${getInputObjectTypeName(input)} {
${fields
  .map(
    field =>
      `  ${field.name}${isRequired(input) ? '' : '?'}: ${getTSType(field)}`,
  )
  .join(EOL)}
}
export type ${getInputObjectTypeFieldsName(input)} =
  | Extract<keyof ${getInputObjectTypeName(input)}, string>
${fields.map(f => `  | { name: '${f.name}', alias?: string  } `).join(EOL)}
  `
}

function renderEnumType(enumType: GraphQLEnumType): string {
  return `\
export type ${getEnumTypeName(enumType)} =
${enumType
  .getValues()
  .map(value => `  | '${value.name}'`)
  .join(EOL)}
  `
}

function getExposableFieldsTypeName(type: GraphQLObjectType) {
  return `${type.name}Fields`
}

function upperFirst(s: string) {
  return s.replace(/^\w/, c => c.toUpperCase())
}

function getTypeFieldArgName(
  type: GraphQLObjectType,
  field: GraphQLField<any, any>,
) {
  return `${type.name}${upperFirst(field.name)}Args`
}

function getObjectTypeFieldsName(type: GraphQLObjectType) {
  return `${type.name}Object`
}

function getInputObjectTypeFieldsName(type: GraphQLInputObjectType) {
  return `${type.name}InputObject`
}

function getObjectTypeFieldsDetailsName(type: GraphQLObjectType) {
  return `${type.name}FieldDetails`
}

function getInputObjectTypeName(type: GraphQLInputObjectType) {
  return `${type.name}`
}

function getEnumTypeName(enumType: GraphQLEnumType): string {
  return `${enumType.name}Values`
}
