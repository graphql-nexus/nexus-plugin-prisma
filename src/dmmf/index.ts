export { get } from './importer'
export { transform } from './transformer'
export { DMMFClass as DMMF } from './DMMFClass'
export { ExternalDMMF as External } from './types'
