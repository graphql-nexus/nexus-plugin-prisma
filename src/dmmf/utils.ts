import * as Photon from '@prisma/photon'
import { transform } from './transformer'
import { DMMFClass } from './DMMFClass'

export function fromPhotonDMMF(photonDMMF: Photon.DMMF.Document): DMMFClass {
  return new DMMFClass(transform(photonDMMF))
}

export const get = (photonClientPackagePath: string): DMMFClass => {
  let photonClientPackage
  try {
    photonClientPackage = require(photonClientPackagePath)
  } catch (error) {
    throw new Error(
      `Could not find photon package at ${photonClientPackagePath}. Check that you have configured your Photon generator block in schema.prisma correctly and run prisma generate.`,
    )
  }
  return fromPhotonDMMF(photonClientPackage.dmmf)
}
