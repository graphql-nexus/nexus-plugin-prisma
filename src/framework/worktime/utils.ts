import * as Prisma from '@prisma/sdk'
import chalk from 'chalk'
import { WorktimeLens } from 'nexus/plugin'
import stripAnsi from 'strip-ansi'
import * as Scaffolders from './scaffolders'

/**
 * Pinned query-engine version. Calculated at build time and based on `@prisma/cli` version
 */
export const PRISMA_QUERY_ENGINE_VERSION: string = require('@prisma/cli/package.json').prisma.version

/**
 * Get the declared generator blocks in the user's PSL file.
 *
 * If Prisma SDK throws a no-models-found error then it is caught, a model is
 * scaffolded, and this function is re-run.
 */
export async function getGenerators(nexus: WorktimeLens, schemaPath: string): Promise<Prisma.Generator[]> {
  return Prisma.getGenerators({
    schemaPath,
    printDownloadProgress: false,
    version: PRISMA_QUERY_ENGINE_VERSION,
  })
    .catch(async (e: Error) => {
      if (isNoModelsDefinedError(e)) {
        await Scaffolders.exampleModelBlock(schemaPath)
        nexus.log.warn(
          `An example model has been scaffolded for you in ${chalk.bold(
            nexus.layout.projectRelative(schemaPath)
          )}`
        )
        return getGenerators(nexus, schemaPath)
      }
      return Promise.reject(e)
    })
    .then(async (generators) => {
      const hasClientGenerator = generators.find((g) => g.options?.generator.provider === 'prisma-client-js')
      if (!hasClientGenerator) {
        await Scaffolders.prismaClientGeneratorBlock(schemaPath)
        nexus.log.warn(
          `A Prisma Client generator block has been scaffolded for you in ${chalk.bold(
            nexus.layout.projectRelative(schemaPath)
          )}`
        )
        // TODO: Generate it programmatically instead for performance reason
        return getGenerators(nexus, schemaPath)
      } else {
        return generators
      }
    })
}

/**
 * Check if the thrown error is of NoModelsDefined type.
 */
function isNoModelsDefinedError(e: Error): boolean {
  return Boolean(stripAnsi(e.message).match(/.*You don't have any models defined.*/))
}
