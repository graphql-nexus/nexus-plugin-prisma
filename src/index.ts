import * as Builder from './builder'

export type NexusPrismaParams = Builder.Options
export const nexusPrismaPlugin = Builder.build
