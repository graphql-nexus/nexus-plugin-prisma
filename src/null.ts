import { DmmfDocument, DmmfTypes } from './dmmf'

/**
 * Take the incoming GraphQL args of a resolver and replaces all `null` values
 * that maps to a non-nullable field in the Prisma Schema, by `undefined` values.
 * 
 * In Prisma, a `null` value has a specific meaning for the underlying database.
 * Therefore, `undefined` is used instead to express the optionality of a field.
 * 
 * In GraphQL however, no difference is made between `null` and `undefined`.
 * This is the reason why we need to convert all `null` values that were assigned to `non-nullable` fields to `undefined`.
 */
export function transformNullsToUndefined(
  args: Record<string, any>,
  schemaArgs: Record<string, DmmfTypes.SchemaArg>,
  dmmf: DmmfDocument,
) {
  const keys = Object.keys(args)
  for (const key of keys) {
    const val = args[key]
    const schemaArg = schemaArgs[key]

    if (!schemaArg) {
      throw new Error(`Could not find schema arg with name: ${key}`)
    }

    const shouldConvertNullToUndefined =
      val === null && schemaArg.inputType.isNullable === false

    if (shouldConvertNullToUndefined) {
      args[key] = undefined
    } else if (isObject(val)) {
      const newSchemaArgs = dmmf.getInputTypeWithIndexedFields(
        schemaArg.inputType.type,
      ).fields

      args[key] = transformNullsToUndefined(args[key], newSchemaArgs, dmmf)
    }
  }

  return args
}

function isObject(obj: any): boolean {
  return obj && typeof obj === 'object'
}
