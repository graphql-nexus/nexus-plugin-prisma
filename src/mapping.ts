import * as DMMF from './dmmf'
import {
  defaultFieldNamingStrategy,
  FieldNamingStrategy,
  OperationName,
} from './naming-strategies'
import { flatMap } from './utils'

interface BaseMappedField {
  field: string
  operation: OperationName
  model: string
  photonAccessor: string
}

export interface MappedField extends Omit<BaseMappedField, 'field'> {
  field: DMMF.Data.SchemaField
}

const buildField = (
  mapping: DMMF.Data.Mapping,
  operation: OperationName,
): BaseMappedField => ({
  operation,
  field: mapping[operation],
  model: mapping.model,
  photonAccessor: mapping.plural,
})

const CRUD_MAPPED_FIELDS: Record<
  string,
  (m: DMMF.Data.Mapping) => BaseMappedField[]
> = {
  Query: m => [buildField(m, 'findOne'), buildField(m, 'findMany')],
  Mutation: m => [
    buildField(m, 'create'),
    buildField(m, 'update'),
    buildField(m, 'updateMany'),
    buildField(m, 'delete'),
    buildField(m, 'deleteMany'),
    buildField(m, 'upsert'),
  ],
}

export const getCrudMappedFields = (
  typeName: 'Query' | 'Mutation',
  dmmf: DMMF.DMMF,
  namingStrategy: FieldNamingStrategy = defaultFieldNamingStrategy,
): MappedField[] =>
  flatMap(dmmf.mappings, m => CRUD_MAPPED_FIELDS[typeName](m)).map(
    mappedField => ({
      ...mappedField,
      field: {
        ...dmmf.getOutputType(typeName).getField(mappedField.field),
        name: namingStrategy[mappedField.operation](
          mappedField.field,
          mappedField.model,
        ),
      },
    }),
  )
