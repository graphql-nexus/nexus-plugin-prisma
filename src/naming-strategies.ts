import camelCase from 'camelcase'
import pluralize from 'pluralize'
import { InternalDMMF } from './dmmf'
import { upperFirst } from './utils'

export interface ArgsNamingStrategy {
  whereInput: (typeName: string, fieldName: string) => string
  orderByInput: (typeName: string, fieldName: string) => string
  relationFilterInput: (typeName: string, fieldName: string) => string
}

export const defaultArgsNamingStrategy: ArgsNamingStrategy = {
  whereInput(typeName, fieldName) {
    return `${upperFirst(typeName)}${upperFirst(fieldName)}WhereInput`
  },
  orderByInput(typeName, fieldName) {
    return `${upperFirst(typeName)}${upperFirst(fieldName)}OrderByInput`
  },
  relationFilterInput(typeName, fieldName) {
    return `${upperFirst(typeName)}${upperFirst(fieldName)}Filter`
  },
}

export type OperationName = Exclude<keyof InternalDMMF.Mapping, 'model' | 'plural'>

export type FieldNamingStrategy = Record<OperationName, (fieldName: string, modelName: string) => string>

export const defaultFieldNamingStrategy: FieldNamingStrategy = {
  // findFirst: (_, modelName) => camelCase(modelName),
  aggregate: (fieldName) => fieldName,
  findUnique: (_, modelName) => camelCase(modelName),
  findMany: (_, modelName) => camelCase(pluralize(modelName)),
  groupBy: (fieldName) => fieldName,
  create: (fieldName) => fieldName,
  createMany: (fieldName) => fieldName,
  delete: (fieldName) => fieldName,
  deleteMany: (fieldName) => fieldName,
  update: (fieldName) => fieldName,
  updateMany: (fieldName) => fieldName,
  upsert: (fieldName) => fieldName,
}
