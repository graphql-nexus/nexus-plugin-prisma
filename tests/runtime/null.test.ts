import { DmmfDocument, InternalDMMF } from '../../src/dmmf'
import { getCrudMappedFields } from '../../src/mapping'
import { OperationName } from '../../src/naming-strategies'
import { transformNullsToUndefined } from '../../src/null'
import { indexBy } from '../../src/utils'
import { getDmmf } from '../__utils'

const operationToRoot: Record<OperationName, 'Query' | 'Mutation'> = {
  findUnique: 'Query',
  findMany: 'Query',
  create: 'Mutation',
  delete: 'Mutation',
  deleteMany: 'Mutation',
  update: 'Mutation',
  updateMany: 'Mutation',
  upsert: 'Mutation',
  aggregate: 'Query',
  createMany: 'Mutation',
  groupBy: 'Query',
}

async function getSchemaArgsForCrud(
  datamodel: string,
  model: string,
  operation: OperationName
): Promise<{
  schemaArgs: Record<string, InternalDMMF.SchemaArg>
  dmmf: DmmfDocument
}> {
  const dmmf = await getDmmf(datamodel)
  const mappedField = getCrudMappedFields(operationToRoot[operation], dmmf).find(
    (x) => x.operation === operation && x.model === model
  )

  if (!mappedField) {
    throw new Error(`Could not find mapped fields for model ${model} and operation ${operation}`)
  }

  return {
    schemaArgs: indexBy(mappedField.field.args, 'name'),
    dmmf,
  }
}

test('findMany: converts nulls to undefined when fields are not nullable', async () => {
  const datamodel = `
    model User {
      id        String   @default(cuid()) @id
      email     String?  @unique
      birthDate DateTime
      posts     Post[]   @relation("PostAuthors")
    }
    
    model Post {
      id      String @default(cuid()) @id
      authors User[] @relation("PostAuthors")
    }
  `
  const { dmmf, schemaArgs } = await getSchemaArgsForCrud(datamodel, 'User', 'findMany')
  const incomingArgs = {
    before: null, // not nullable
    after: null, // not nullable
    first: 1,
    orderBy: {
      email: null, // nullable
      birthDate: null, // nullable
      id: 'asc',
    },
    where: {
      AND: null, // not nullable
      NOT: {
        AND: {
          birthDate: null, // not nullable
        },
        posts: null,
      },
    },
  }

  const result = transformNullsToUndefined(incomingArgs, schemaArgs, dmmf)

  expect(result).toMatchSnapshot()
})

test('create: converts nulls to undefined when fields are not nullable', async () => {
  const datamodel = `
  model User {
    id        String   @default(cuid()) @id
    email     String?  @unique
    birthDate DateTime
    posts     Post[]   @relation("PostAuthors")
  }
  
  model Post {
    id      String @default(cuid()) @id
    authors User[] @relation("PostAuthors")
  }
  `
  const { dmmf, schemaArgs } = await getSchemaArgsForCrud(datamodel, 'User', 'create')
  const incomingArgs = {
    data: {
      id: null, // not nullable
      posts: {
        connect: null, // not nullable
        create: {
          id: 'titi',
        },
      },
    },
  }

  const result = transformNullsToUndefined(incomingArgs, schemaArgs, dmmf)

  expect(result).toMatchSnapshot()
})

test('model filtering: converts nulls to undefined when fields are not nullable', async () => {
  const datamodel = `
  model User {
    id        String   @default(cuid()) @id
    email     String?  @unique
    birthDate DateTime
    posts     Post[]   @relation("PostAuthors")
  }
  
  model Post {
    id      String @default(cuid()) @id
    authors User[] @relation("PostAuthors")
  }
  `
  const dmmf = await getDmmf(datamodel)
  const schemaArgs = dmmf.getOutputType('User').fields.find((f) => f.name === 'posts')?.args!
  const indexedSchemaArgs = indexBy(schemaArgs, 'name')
  const incomingArgs = {
    where: {
      id: null, // not nullable
      authors: {
        every: {
          email: null,
          birthDate: null, // not nullable
          posts: null,
        },
      },
    },
  }

  const result = transformNullsToUndefined(incomingArgs, indexedSchemaArgs, dmmf)

  expect(result).toMatchSnapshot()
})

test('do not convert args that are arrays', async () => {
  const datamodel = `
  model User {
    id        String   @default(cuid()) @id
    email     String?  @unique
    birthDate DateTime
    posts     Post[]   @relation("PostAuthors")
  }
  
  model Post {
    id      String @default(cuid()) @id
    authors User[] @relation("PostAuthors")
  }
  `
  const dmmf = await getDmmf(datamodel)
  const schemaArgs = dmmf.getOutputType('User').fields.find((f) => f.name === 'posts')?.args!
  const indexedSchemaArgs = indexBy(schemaArgs, 'name')
  const incomingArgs = {
    where: {
      OR: [{ something: true }, { something: false }],
    },
  }

  const result = transformNullsToUndefined(incomingArgs, indexedSchemaArgs, dmmf)

  expect(result).toMatchSnapshot()
})

test('safely transforms json fields', async () => {
  const datamodel = `
  datasource db {
    provider = "postgresql"
    url      = "postgresql://"
  }

  model User {
    id  Int @id @default(autoincrement())
    json1 Json
    json2 Json
    optionalJson1 Json?
    optionalJson2 Json?
  }
  `
  const { dmmf, schemaArgs } = await getSchemaArgsForCrud(datamodel, 'User', 'create')

  const incomingArgs = {
    data: {
      json1: null, // not nullable
      json2: { someRandomJson: 'titi' },
      optionalJson1: null,
      optionalJson2: { someRandomJson: 'titi' },
    },
  }

  const result = transformNullsToUndefined(incomingArgs, schemaArgs, dmmf)

  expect(result).toMatchInlineSnapshot(`
    Object {
      "data": Object {
        "json1": undefined,
        "json2": Object {
          "someRandomJson": "titi",
        },
        "optionalJson1": null,
        "optionalJson2": Object {
          "someRandomJson": "titi",
        },
      },
    }
  `)
})
