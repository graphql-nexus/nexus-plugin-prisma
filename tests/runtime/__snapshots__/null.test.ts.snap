// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create: converts nulls to undefined when fields are not nullable 1`] = `
{
  "data": {
    "id": undefined,
    "posts": {
      "connect": undefined,
      "create": {
        "id": "titi",
      },
    },
  },
}
`;

exports[`do not convert args that are arrays 1`] = `
{
  "where": {
    "OR": [
      {
        "something": true,
      },
      {
        "something": false,
      },
    ],
  },
}
`;

exports[`findMany: converts nulls to undefined when fields are not nullable 1`] = `
{
  "after": undefined,
  "before": undefined,
  "first": 1,
  "orderBy": {
    "birthDate": undefined,
    "email": undefined,
    "id": "asc",
  },
  "where": {
    "AND": undefined,
    "NOT": {
      "AND": {
        "birthDate": undefined,
      },
      "posts": undefined,
    },
  },
}
`;

exports[`model filtering: converts nulls to undefined when fields are not nullable 1`] = `
{
  "where": {
    "authors": {
      "every": {
        "birthDate": undefined,
        "email": null,
        "posts": undefined,
      },
    },
    "id": undefined,
  },
}
`;
